File: lib/server/account.lua
---@diagnostic disable: redundant-parameter
---@class OxAccountServer : OxClass
local OxAccount = lib.class('OxAccount')

function OxAccount:__index(index)
    local value = OxAccount[index] --[[@as any]]

    if type(value) == 'function' then
        self[index] = value == OxAccount.__call and function(...)
            return value(self, index, ...)
        end or function(...)
            return value(self, ...)
        end

        return self[index]
    end

    return value
end

function OxAccount:constructor(data)
    for k, v in pairs(data) do
        self[k] = v
    end
end

function OxAccount:__call(...)
    return exports.ox_core:CallAccount(self.accountId, ...)
end

function OxAccount:__tostring()
    return json.encode(self, { indent = true })
end

for method in pairs(exports.ox_core:GetAccountCalls() or {}) do
    if not rawget(OxAccount, method) then OxAccount[method] = OxAccount.__call end
end

local function CreateAccountInstance(account)
    if not account then return end

    return OxAccount:new(account)
end

---@class OxServer
local Ox = Ox

function Ox.GetAccount(accountId)
    local account = exports.ox_core:GetAccount(accountId)
    return CreateAccountInstance(account)
end

function Ox.GetCharacterAccount(charId)
    local account = exports.ox_core:GetCharacterAccount(charId)
    return CreateAccountInstance(account)
end

function Ox.GetGroupAccount(groupName)
    local account = exports.ox_core:GetGroupAccount(groupName)
    return CreateAccountInstance(account)
end

function Ox.CreateAccount(owner, label)
    local account = exports.ox_core:CreateAccount(owner, label)
    return CreateAccountInstance(account)
end

File: lib/server/account.ts
import type { OxAccount as _OxAccount } from 'server/accounts/class';

class AccountInterface {
  constructor(public accountId: number) {}
}

Object.keys(exports.ox_core.GetAccountCalls()).forEach((method: string) => {
  (AccountInterface.prototype as any)[method] = function (...args: any[]) {
    return exports.ox_core.CallAccount(this.accountId, method, ...args);
  };
});

AccountInterface.prototype.toString = function () {
  return JSON.stringify(this, null, 2);
};

export type OxAccount = _OxAccount & AccountInterface;

function CreateAccountInstance(account?: _OxAccount) {
  if (!account) return;

  return new AccountInterface(account.accountId) as OxAccount;
}

export async function GetAccount(accountId: number) {
  const account = await exports.ox_core.GetAccount(accountId);
  return CreateAccountInstance(account);
}

export async function GetCharacterAccount(charId: number | string) {
  const account = await exports.ox_core.GetCharacterAccount(charId);
  return CreateAccountInstance(account);
}

export async function GetGroupAccount(groupName: string) {
  const account = await exports.ox_core.GetGroupAccount(groupName);
  return CreateAccountInstance(account);
}

export async function CreateAccount(owner: number | string, label: string) {
  const account = await exports.ox_core.CreateAccount(owner, label);
  return CreateAccountInstance(account);
}
